pkg/controller/instance/instance_controller.go

```diff
@@ -543,19 +516,112 @@ func isUpgradePlan(planName string) bool {
+		if plan, err = fetchNewExecutionPlan(i, ov); plan != nil && err == nil {
+			log.Printf("InstanceController: Fetched new execution plan '%s' from the spec for instance %s/%s", kudo.StringValue(plan), i.Namespace, i.Name)
+		}
+	} else {
+		if plan, err = inferNewExecutionPlan(i, ov); plan != nil && err == nil {
+			log.Printf("InstanceController: Inferred new execution plan '%s' from instance %s/%s state", kudo.StringValue(plan), i.Namespace, i.Name)
+		}
+	}
+
+	return
+}
+
+// fetchNewExecutionPlan method fetches a new execution plan from Instance.Spec.PlanExecution field. A plan is actually
+// new if the PlanExecution.UID has changed, otherwise it's just an old plan in progress.
+// - "planName", when there is a new plan that needs to be executed
+// - <nil>, no new plan found e.g. a plan is already in progress
+func fetchNewExecutionPlan(i *v1beta1.Instance, ov *v1beta1.OperatorVersion) (*string, error) {
+	snapshot, err := i.SnapshotSpec()
+	if err != nil {
+		return nil, &v1beta1.InstanceError{Err: fmt.Errorf("failed to unmarshal instance snapshot %s/%s: %v", i.Namespace, i.Name, err), EventName: kudo.String("UnexpectedState")}
+	}
+
+	// Instance deletion is an edge case where the admission webhook *can not* populate the Spec.PlanExecution.PlanName
+	// with the 'cleanup' plan. So we have to do it here ourselves. Only if:
+	// 1. Instance is being deleted
+	// 2. Cleanup plan exists in the operator version and has *never run* before
+	// 3. Cleanup hasn't been scheduled yet (first time the deletion is being reconciled)
+	// we set the Spec.PlanExecution.PlanName = 'cleanup'
+	hasToScheduleCleanupAfterDeletion := func() bool {
+		shouldCleanup := i.IsDeleting() && kudov1beta1.CleanupPlanExists(ov)
+		cleanupNeverRun := i.PlanStatus(v1beta1.CleanupPlanName) == nil || i.PlanStatus(v1beta1.CleanupPlanName).Status == kudov1beta1.ExecutionNeverRun
+		cleanupNotScheduled := i.Spec.PlanExecution.PlanName != v1beta1.CleanupPlanName
+
+		return shouldCleanup && cleanupNeverRun && cleanupNotScheduled
+	}
+	if hasToScheduleCleanupAfterDeletion() {
+		log.Printf("InstanceController: Instance %s/%s is being deleted. Scheduling '%s' plan.", i.Namespace, i.Name, v1beta1.CleanupPlanName)
+
+		i.Spec.PlanExecution.PlanName = v1beta1.CleanupPlanName
+		i.Spec.PlanExecution.UID = uuid.NewUUID()
+	}
+
+	newPlanScheduled := func() bool {
+		// this can happen if the user installed an instance without an active webhook (e.g. in the previous KUDO version)
+		// as long as we don't have an upgrade path, we must handle this
+		if snapshot == nil {
+			return i.Spec.PlanExecution.PlanName != ""
+		}
+		oldPlan := snapshot.PlanExecution.PlanName
+		oldUID := snapshot.PlanExecution.UID
+		newPlan := i.Spec.PlanExecution.PlanName
+		newUID := i.Spec.PlanExecution.UID
+
+		return (newPlan != oldPlan && newPlan != "") || // either there is a non-empty new plan
+			(oldPlan != "" && newPlan == oldPlan && newUID != oldUID) // or the same plan with a new UID
```

Comment:
```suggestion
                 // A non-empty new plan is set, and it's not the old plan or the UID has changed
		return newPlan != "" && ((newPlan != oldPlan || newUID != oldUID )
```

Did I miss a case there? I'm not fully sure why we check `oldPlan != ""`

Reply:
This looks equivalent to me.  In general, I spent so much time fixing subtle bugs in these checks (in the webhook) so I started being explicit (even unnecessary) and optimize for readability.  Reminds me that I wanted to write some tests for this method ;)